### DISCUSSION TITLE
如何在多个 project 中复用函数

### DISCUSSION THREAD

**Discussion by Dont-laugh** [xmake-io/xmake#4779] [Q&A]
我在一个 Solution 中有多个 Projects，每个 Project 都对应一个 xmake.lua 配置，目录大致如下：
```
.
xmake.lua
├── project_1
|  └── xmake.lua
|  └── resource/
├── project_2
|  └── xmake.lua
|  └── resource/
```

在根目录下的 xmake.lua 中一个个 includes 其他 project 目录，由于每个 project 都有一些自己的资源，我需要在 on_build 的时候把资源拷贝到执行目录，它们完全用同一个函数来执行拷贝逻辑。
我尝试过在根目录的 xmake.lua 中定义这个拷贝函数，子 xmake 中会找不到这个函数。但神奇的是，我在根 xmake 中定义的变量却可以在子 xmake 中找到，在 on_build 中要用到这个变量也只需要在 target 范围外用一个 local 变量存一下就行，函数就算存下来任然是 nil。
所以，想要在多个 xmake.lua 中复用某个函数的话，我应该怎么做？

``` lua
# ./xmake.lua

function copy_resouces(proj_name, target)
    local src_dir = format("%s/resource/*", proj_name)
    local target_dir = target:targetdir()
    for _, file in ipairs(os.files(src_dir)) do
        os.cp(file, path.join(target_dir, proj_name, path.filename(file)))
    end
end

var_test = 233
includes("project_1")

# project_1/xmake.lua

local upper_var = var_test

target("project_1")
    set_kind("binary")
    add_files("*.cpp")
    on_build(function(target)
        print(upper_var)                    # 233
        copy_resouces("project_1", target)  # 此时会报空
    end)
target_end()
```

**Comment by waruqi** [comment]
描述域和脚本域是隔离的，不要混用。。脚本域里面 走 import 导入模块使用，用户自定义模块也可以 import 复用，add_moduledirs 配置模块目录，具体看下文档

**Comment by star-hengxing** [comment]
自动义脚本写好后用 [import](https://xmake.io/#/manual/builtin_modules?id=import)。
写一个 rule，在 `after_build` 执行这个脚本。
在全局 `add_rules`，这样每个 target 都能执行

> 资源文件建议用 `os.ln`，避免拷贝
